pepper-apis/docs/specification/src/components/schemas.yml

```diff
@@ -2894,3 +2933,49 @@ Suggestion.ActivityInstance:
       description: Activity Instance Name, rendered, as could be seen on the dashboard
       type: string
       example: "Medical Record Release Form #2"
+DateRange:
+  type: object
+  required:
+    - from
+    - to
+  properties:
+    from:
+      type: string
+      description: the lower constraint of the range. This value is inclusive. The date format is a full, extended ISO-8601 Calendar date and time in the format `YYYY-MM-DDTHH:MM:SSZ` in the UTC time zone.
+      nullable: true
+      format: date-time
+    to:
+      type: string
+      description: the upper constraint of the range. This value is inclusive. The date format is a full, extended ISO-8601 Calendar date and time in the format `YYYY-MM-DDTHH:MM:SSZ` in the UTC time zone.
+      nullable: true
+      format: date-time
+IntegerRange:
+  type: object
+  required:
+    - from
+    - to
+  properties:
+    from:
+      type: integer
+      description: the lower constraint of the range. This value is inclusive.
+      nullable: true
+      format: int64
+    to:
+      type: integer
+      description: the upper constraint of the range. This value is inclusive.
+      nullable: true
+      format: int64
```

Comment:
I don't think the from/to is enough to enable pagination.
Client is likely to have a limit of items that wants to display at a time, but does not know how many total rows there will be in a particular result set.
So I think that one capability we would want is for client to know on the first query what is the total number of results so it it know if it should allow user to request another page of results or not.
This could be achieved by including a "limit" in the query and returning a "total" in the results. 
A follow up query then could specify an offset/index to start reading the next set of values (in first query, the offset default value is "0").
(I am not saying that we should do exactly what I wrote, just suggesting that this is the sort of interaction that would be needed with client).
Another problem that would be SUPER NICE if we dealt with it in API and implementation is if we can keep the result set "locked" during pagination.
What I mean is that the audit trail data set will continually grow/always new rows are going to be added and can be a little confusing for users to do an initial query, find that there are 350 results, click for the next page and find out that there are 352 results now and that the second page includes some of the same results from the first page.
There are multiple strategies that can be used to keep this dataset locked while paging.
I recommend doing some Googling to see how other folks have resolved this REST API paging problem and see what we can learn from prior art.
A reminder: this audit table is going to become very quickly the largest in Pepper! So this paging piece is super-important. But no pressure!

Reply:
Well, a couple of comments from my side.
1. Regarding the pagination: I'm not quite sure I follow you entirely. When you implement infinite scroll, once the user reaches the end of the page FE sends another query with the shifted start index. And this happens until BE returns not empty result. To avoid delays in rendering, we can use "double buffering" technics and preload the next page in advance.
2. As for the result set, the simplest thing we can do here is to create temporary table or view and assign a hash to each view. Then, based on the hash we can fetch and narrow down the result set from the view. So, filter will be extended by result set hash.

Reply:
But do we really need this? I mean most likely user will request the logs in natural order (id ASC or date_created ASC). So the situation you described wouldn't be possible, right?

Reply:
OK. I don't think we will be doing infinite scrolling. We will want to do an initial query and the client will want to know if there are more results. And if there are more results how many there are.
Ideally, we client will be able to calculate how many pages of results there are to display (maybe we will have a Next and a Previous button or maybe a list of buttons with the page numbers).
One of the things that I am requesting we enable in the design of the API (and later, the implementation), is that, after the initial query, if the client goes to page 1 of results and the page 2 and 3, if they go back to page 2 they will see the exact same results they had seen before on page 2 UNLESS they do a brand new query.

Reply:
The default order should actually be DESC by time stamp: the most recent events at the top. But even if sorting is not required (we need to check with David O on sorting), you can see that if you go back to the first page of results AND a new event has occurred since the query, there can be potentially new items that were not there before.

Reply:
I think this dude had a good discussion of the issues involved in designing this sort of APIs.
https://www.moesif.com/blog/technical/api-design/REST-API-Design-Filtering-Sorting-and-Pagination/

Reply:
Agreed with DESC by default.
I went through the article you sent me and as far as I understood, what we need is [Keyset pagination](https://www.moesif.com/blog/technical/api-design/REST-API-Design-Filtering-Sorting-and-Pagination/#keyset-pagination). And it is totally possible to implement using current API schema by the combination of `limitation` statement and `dateRange` restriction.

Reply:
Is IntegerRange at line 2893 being used anymore?

Reply:
Still not seeing how the client will know how many pages of results are available after initial query. Don't we need a count of the total number of results available? How will client know that there are more results?

Reply:
Probably we need to implement one more method with the same signature - **count**. Accepting the filter will return the only number - total results matching the criteria.

I was a bit confused by pagination since it is quite an outdated approach as far as I know.

Answering your question: how the client will know if there are more results:
Front-end performs two parallel requests:
1. Search(0; 10)
2. Search(10, 10)

Front-end renders only the first one. Once the page is scrolled to the bottom it executes the next search (20; 10) and so on. Until the number of results is less than the requested count. Once we received fewer results than queried that means it was the last batch.

It is some kind of double-buffering. So the data will be loaded smoothly.

Reply:
If we still want to support the pages, then my suggestion is to implement pair of the methods:
1. List<R> search(F filter) - does the real search
2. Integer count(F filter) - count the number of results matching the filter (ignoring the Pagination property)

Having those two UI approaches might be implemented: infinite scroll and pagination. Pagination might be reached by calling count just once and dividing the number of the results by page size.